libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -1161,25 +1206,53 @@ void WaveClip::ShiftBy(double delta) noexcept
 
 // Bug 2288 allowed overlapping clips.
 // This was a classic fencepost error.
-// We are within the clip if start < t <= end.
+// We are within the clip if start <= t < end.
 // Note that BeforeClip and AfterClip must be consistent
 // with this definition.
 bool WaveClip::WithinPlayRegion(double t) const
 {
-    auto ts = TimeToSamples(t);
-    return ts > GetPlayStartSample() && ts < GetPlayEndSample() + GetAppendBufferLen();
+   return GetPlayStartTime() <= t && t < GetPlayEndTime();
+}
+
+bool WaveClip::EntirelyWithinPlayRegion(double t0, double t1) const
+{
+   assert(t0 <= t1);
+   return !BeforePlayRegion(t0) && !AfterPlayRegion(t1);
+}
+
+bool WaveClip::PartlyWithinPlayRegion(double t0, double t1) const
+{
+   assert(t0 <= t1);
+   return OverlapsPlayRegion(t0, t1) && !EntirelyWithinPlayRegion(t0, t1);
+}
+
+bool WaveClip::OverlapsPlayRegion(double t0, double t1) const
+{
+   assert(t0 <= t1);
+   return !(BeforePlayRegion(t1) || AfterPlayRegion(t0));
+}
+
+bool WaveClip::ExtendsPlayRegion(double t0, double t1) const
+{
+   assert(t0 <= t1);
+   return (BeforePlayRegion(t0) && WithinPlayRegion(t1)) ||
+          (WithinPlayRegion(t0) && AfterPlayRegion(t1));
+}
+
+bool WaveClip::ExtendsPlayRegionOnBothSides(double t0, double t1) const
```

Comment:
Or simply "IncludesPlayRegion"?

Reply:
That name is unsatisfactory -- rather "PlayRegionWithin" ?

Its the interval that extends the play region, not the opposite.  When I read a member function name, I think of `this` as the subject of whatever verb is in it.

All right, some better naming is needed for this set of functions, and I haven't necessarily given the best names yet.

Reply:
>  I think of this as the subject of whatever verb is in it.

Interesting to observe is that `BeforePlayRegion(double t)` and other methods taking a time point, this ambiguity doesn't seem to arise. I guess `ExtendsPlayRegionOnBothSides` has this problem because `t0` and `t1` may be interpreted as a play region.
Renaming `ExtendsPlayRegionOnBothSides` to `PlayRegionIsExtendedOnBothSidesBy(t0, t1)` would require also renaming `BeforePlayRegion` and the likes to `PlayRegionIsAfter` ; do we really want this? I think getting used to `ExtendsPlayRegionOnBothSides` is easy.

Reply:
That said, I'm ok to rename `ExtendsPlayRegionOnBothSides` to `PlayRegionStrictlyWithin`.